eigensync/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+use std::{collections::HashMap, fmt::Debug, marker::PhantomData, sync::{Arc, OnceLock}, time::Duration};
+pub mod protocol;
+pub mod database;
+
+use anyhow::{bail, Context};
+//pub mod protocol;
+use libp2p::{
+    futures::StreamExt, 
+    identity, noise, request_response::{self, OutboundRequestId}, 
+    swarm::SwarmEvent, 
+    tcp, yamux, Multiaddr, PeerId, Swarm, SwarmBuilder
+};
+use tokio::{sync::{mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender}, oneshot, RwLock}, task};
+//use crate::protocol::Request;
+use automerge::{ActorId, AutoCommit, Change};
+use autosurgeon::{hydrate, reconcile, Hydrate, Reconcile};
+use tokio_util::task::AbortOnDropHandle;
+use crate::protocol::{client, Behaviour, BehaviourEvent, ChannelRequest, Response, SerializedChange, SignedEncryptedSerializedChange, ServerRequest};
+
+pub type SyncBehaviour = request_response::cbor::Behaviour<ServerRequest, Response>;
+
+pub struct SyncLoop {
+    receiver: UnboundedReceiver<ChannelRequest>,
+    response_map: HashMap<OutboundRequestId, oneshot::Sender<Result<Vec<SignedEncryptedSerializedChange>, String>>>,
+    swarm: Swarm<Behaviour>,
+    connection_established: Option<oneshot::Sender<()>>,
+}
+
+impl SyncLoop {
+    pub async fn new(receiver: UnboundedReceiver<ChannelRequest>, swarm: Swarm<Behaviour>, connection_established: oneshot::Sender<()>) -> Self {
+        
+        Self { receiver, response_map: HashMap::new(), swarm, connection_established: Some(connection_established)}
+    }
+    
+    pub fn sync_with_server(&mut self, request: ChannelRequest, server_id: PeerId) {
+        let server_request = ServerRequest::UploadChangesToServer { encrypted_changes: request.encrypted_changes };
+        match &server_request {
```

Comment:
We do not need to match here because `server_request` is always of type `ServerRequest::UploadChangesToServer`